lang/py/src/avro/schema.py

```diff
@@ -405,6 +420,106 @@ def __eq__(self, that):
     return self.props == that.props
 
 #
+# Decimal Type
+#
+class DecimalSchema(LogicalSchema):
+  def __init__(self, decimal_type, precision, scale):
+    LogicalSchema.__init__(self, decimal_type, 'decimal')
+    self.set_prop('precision', precision)
+    self.set_prop('scale', scale)
+
+  def validate(self, validation_props=None):
+    if not isinstance(validation_props.get('precision'), int):
+      raise SchemaParseException('Precision (Integer) is required for logical type DECIMAL.')
+
+    precision = validation_props.get('precision')
+    if precision <= 0:
+      raise SchemaParseException("Invalid DECIMAL precision %s. Must be positive."
+                                 % precision)
+    elif precision > validation_props.get('max_precision'):
+      raise SchemaParseException("Fixed %s cannot store precision digits. Max is %s"
+                                 %(validation_props.get('size'),
+                                   validation_props.get('max_precision')))
+
+    if not isinstance(validation_props.get('scale'), int):
+      raise SchemaParseException('Scale must be of type Integer.')
+
+    scale = validation_props.get('scale')
+    if scale < 0:
+      raise SchemaParseException("Invalid DECIMAL scale %s. Must be positive"
+                                 % scale)
+    elif scale > precision:
+      raise SchemaParseException("Invalid DECIMAL scale %s. Cannot be greater than precision %s"
+                                 %(scale, precision))
+
+#
+# Decimal Bytes Type
+#
+
+class BytesDecimalSchema(DecimalSchema):
+  def __init__(self, schema_type, precision, scale):
+    if schema_type != 'bytes':
+      raise SchemaParseException('Logical type DECIMAL must be backed by fixed or bytes.')
+
+    self.validate({'max_precision': self._max_precision(),
+                   'precision': precision,
+                   'scale': scale})
+
+    DecimalSchema.__init__(self, schema_type, precision, scale)
+
+  def _max_precision(self):
+    # Considering the max 32 bit integer value
+    return (1 << 31) - 1
+
+  def to_json(self, names=None):
+    if len(self.props) == 1:
```

Comment:
I don't think this branch can ever be taken (since all decimals have at least a `type` and a `precision`).